target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl931x.c

```diff
@@ -697,33 +696,38 @@ static void rtl931x_fill_l2_row(u32 r[], struct rtl838x_l2_entry *e)
 		return;
 	}
 
-	r[2] = BIT(31);	/* Set valid bit */
+	r[0] = BIT(31); /* Set valid bit */
 
-	r[0] = ((u32)e->mac[0]) << 24 |
-	       ((u32)e->mac[1]) << 16 |
-	       ((u32)e->mac[2]) << 8 |
-	       ((u32)e->mac[3]);
-	r[1] = ((u32)e->mac[4]) << 24 |
-	       ((u32)e->mac[5]) << 16;
+	r[0] |= ((u32)e->mac[0]) << 8 |
+	       ((u32)e->mac[1]);
+	r[1] = ((u32)e->mac[2]) << 24 |
+	       ((u32)e->mac[3]) << 16 |
+		   ((u32)e->mac[4]) << 8 |
+		   ((u32)e->mac[5]);
 
-	r[2] |= e->next_hop ? BIT(12) : 0;
+	r[0] |= e->is_open_flow ? BIT(30) : 0;
+	r[0] |= e->is_pe_forward ? BIT(28) : 0;
+	r[2] |= e->next_hop ? BIT(30) : 0;
+	r[0] |= (e->rvid & 0xfff) << 16;
 
 	if (e->type == L2_UNICAST) {
-		r[2] |= e->is_static ? BIT(14) : 0;
-		r[1] |= e->rvid & 0xfff;
-		r[2] |= (e->port & 0x3ff) << 20;
+		r[2] |= e->is_l2_tunnel ? BIT(31) : 0;
+		r[2] |= e->is_static ? BIT(13) : 0;
+		r[2] |= (e->port & 0x3ff) << 19;
+
 		if (e->is_trunk) {
-			r[2] |= BIT(30);
-			port = e->stack_dev << 9 | (e->port & 0x3f);
+			r[2] |= BIT(29);
+			port = e->trunk & 0xff;
 		} else {
-			port = (e->stack_dev & 0xf) << 6;
-			port |= e->port & 0x3f;
+			port = e->port & 0x3f;
+			port |= (e->stack_dev & 0xf) << 6;
 		}
-		r[2] |= port << 20;
-		r[2] |= e->block_da ? BIT(15) : 0;
-		r[2] |= e->block_sa ? BIT(17) : 0;
-		r[2] |= e->suspended ? BIT(13) : 0;
-		r[2] |= (e->age & 0x3) << 17;
+
+		r[2] |= (port & 0x3ff) << 19;
+		r[2] |= e->block_da ? BIT(14) : 0;
+		r[2] |= e->block_sa ? BIT(15) : 0;
+		r[2] |= e->suspended ? BIT(12) : 0;
+		r[2] |= (e->age & 0x3) << 16;
```

Comment:
Age has 3 bits and not only two. Mask should be 0x7. Same for `rtl931x_fill_l2_entry`